packages/smart-contracts/src/contracts/BatchConversionPayments.sol

```diff
@@ -0,0 +1,409 @@
+          uTokens[k].amountAndFee = requestsInfo[i]._maxToSpend;
+          break;
+        }
+      }
+    }
+
+    IERC20 requestedToken;
+    // For each token: check allowance, transfer funds on the contract and approve the paymentProxy to spend if needed
+    for (uint256 k = 0; k < uTokens.length && uTokens[k].amountAndFee > 0; k++) {
+      requestedToken = IERC20(uTokens[k].tokenAddress);
+      uTokens[k].batchFeeAmount = (uTokens[k].amountAndFee * batchConversionFee) / 10000;
+      // Check proxy's allowance from user, and user's funds to pay approximated amounts.
+      require(
+        requestedToken.allowance(msg.sender, address(this)) >= uTokens[k].amountAndFee,
+        'Not sufficient allowance for batch to pay'
+      );
+      require(requestedToken.balanceOf(msg.sender) >= uTokens[k].amountAndFee, 'not enough funds');
+      require(
+        requestedToken.balanceOf(msg.sender) >= uTokens[k].amountAndFee + uTokens[k].batchFeeAmount,
+        'not enough funds to pay approximated batchConversionFee'
+      );
+
+      // Transfer the amount and fee required for the token on the batch conversion contract
+      require(
+        safeTransferFrom(uTokens[k].tokenAddress, address(this), uTokens[k].amountAndFee),
+        'payment transferFrom() failed'
+      );
+
+      // Batch contract approves Erc20ConversionProxy to spend the token
+      if (
+        requestedToken.allowance(address(this), address(conversionPaymentProxy)) <
+        uTokens[k].amountAndFee
+      ) {
+        approvePaymentProxyToSpend(uTokens[k].tokenAddress, address(conversionPaymentProxy));
+      }
+    }
+
+    // Batch pays the requests using Erc20ConversionFeeProxy
+    for (uint256 i = 0; i < requestsInfo.length; i++) {
+      RequestInfo memory rI = requestsInfo[i];
+      conversionPaymentProxy.transferFromWithReferenceAndFee(
+        rI._recipient,
+        rI._requestAmount,
+        rI._path,
+        rI._paymentReference,
+        rI._feeAmount,
+        _feeAddress,
+        rI._maxToSpend,
+        rI._maxRateTimespan
+      );
+    }
+
+    // Batch sends back to the payer the tokens not spent and pays the batch fee
+    for (uint256 k = 0; k < uTokens.length && uTokens[k].amountAndFee > 0; k++) {
+      requestedToken = IERC20(uTokens[k].tokenAddress);
+
+      // Batch sends back to the payer the tokens not spent = excessAmount
+      // excessAmount = maxToSpend - reallySpent, which is equal to the remaining tokens on the contract
+      uint256 excessAmount = requestedToken.balanceOf(address(this));
+      if (excessAmount > 0) {
+        requestedToken.safeTransfer(msg.sender, excessAmount);
+      }
+
+      // Payer pays batch fees amount
+      require(
+        safeTransferFrom(
+          uTokens[k].tokenAddress,
+          _feeAddress,
+          ((((uTokens[k].amountAndFee - excessAmount) * 10000) / (10000 + basicFee)) *
+            batchConversionFee) / 10000
```

Comment:
Just to be sure: we are not using `uTokens[k].batchFeeAmount` that we've computed earlier because it's an estimation, and insead we are computing the fees on the effective amount paid ?

Reply:
exactly, it is only for require checking.
I have added:
 * @dev amountAndFee is an approximation of the amount and the fee to be paid, in order to get enough tokens.
   *                   The excess is sent back to the payer
   *      batchFeeAmount is an approximation for the same reason of amountAndFee